### ISSUE TITLE
GSoC Draft Plan

### ISSUE THREAD

**Issue by 12Chao** [comocheng/ChemCheck#8]
Hi @rwest , @bryanwweber This is a draft plan for the project, I leave last two weeks with no plan because I am not sure how long will it take for solving chemical errors( maybe we need a further discussion on it). I plan to get the website construction done before first evaluation period, and I am not sure if it is satisfied for your expectation, please let me know how do you think about the plan.

PS: I figured out how to handle the list view and working on using ajax to integrate ck2cti file.

- Jun 17-21:
Get the website done, achieve the interaction between the script and web frontend, so that I can focus on  debugging script writting in the rest of time. 
- Jun 24 - 28
Evaluation period. Discuss with mentors about how to solve syntax errors. Start with slolving syntax error, I tend to start with duplicated lines first and missing 'E' error
- Jul 1-5
Solving missing digits from end of line and other syntax error (if found)
- Jul 8-12
Start working on Chemical error, I would like to choose  'exceeding collision limit' as first one because I guess this part of code can be similar to [debugging script ](https://github.com/Cantera/ncm-2017-materials/blob/master/mech_debug/thermo_debugging.ipynb)on catera github.
- Jul 15-19
Cooperate with mentors solving discontinuous polynomials and unconsistent thermo for explicit reverse rates 
- Jul 22-26
Evaluation period. Discuss with mentors about the improvement and plan for next phase.
- Jul 29- Aug 2
Working on simplify CVODES error message and dead-end pathways
- Aug 5-9
Keep working if above works are not completed, doing some improvement to Website UI if in feature
- Aug 12- 16
Keep working if above works are not completed, doing some improvement to Website UI if in feature

**Comment by bryanwweber** [comment]
I have a few comments. Also tagging @kyleniemeyer and @decaluwe for their comments. My main concern is that I want to make sure there is at least one working, useful, feature, instead of 5 half-working, not-useful features. For instance, a site that just displays static, colored, HTML of the spot in a CK file with a syntax error and lets me upload a new file I fixed myself on my computer is more useful to me than one that lets me edit the file online but doesn't show me anything about where the error is.

> Get website done

What does this mean? It is really broad, I could interpret "get website done" as finishing the project! Could you be more specific here?

> solve syntax error

I'm not sure what this means either. Do you mean that you will try to automatically correct the problem? I think that would be too complicated for the first version of this website. I think the best thing to start with is just to have a nice interface to display the syntax error that ck2yaml has found, with coloring if possible. If the user can edit the file right on the website, that's great but not necessary as long as they can see the exact line in the file where the problem is.

Some other general comments:

- Can you please provide a summary/list of what features will be completed before each evaluation period?
- You need to give yourself some time in here to write your blog posts (the next one is due this Friday 😄)
- I think we should schedule weekly or bi-weekly telecons (15-30 minutes) to make sure the project is staying on track (@rwest & @kyleniemeyer thoughts on that?)
- Please post your pull requests into the Slack channel

**Comment by kyleniemeyer** [comment]
I agree with everything @bryanwweber said

**Comment by decaluwe** [comment]
Hi @12Chao -- I think this is a good framework for your work plan, but as @bryanwweber said, needs some more detail.  In short, we should be able to read this work plan and very clearly decide if the timeline and objectives are being met.  

In addition to @bryanwweber's comments, I would add the following:

1. During evaluation weeks, I think there needs to also be parallel work objectives.  I.e. to assume that the entire week would be spent "discuss[ing] with mentors" (as in July 22-26) is not very reasonable.
2. For each objective, it would be great to have a little more detail on your priorities and objectives.  What does it mean to "solve" an error?  Will the checker "fix" the problem for the user, or just diagnose and give feedback/options to the user?  If the latter, specifically what information should it provide?  

I know it requires a good bit of thought, and it's okay if sometimes the answer is "I won't really know until I get there." 😆   But defining the criteria as specifically as possible will help guide your work and make it flow a lot easier, going forward.

Lastly, don't worry too much about getting the timing exactly right - we are allowed to update and edit the scope, as we move forward, so long as sufficient progress is made!

Again, this is a very good start - will look forward to seeing how it looks with a little more detail filled in!

**Comment by 12Chao** [comment]
Thanks for your suggestions @bryanwweber @kyleniemeyer @decaluwe , I just revised the plan, and I will add more details as I move forward, especially for debugging script writing plan.

- Jun 17-21:
Build interaction between ck2cti file and web, users can see their input file with comments returned from ck2cti file on the web. This is the most important part in web construction, and after that I can add debugging script into the interaction which achieves other funcions like highlighting. I am using jQuery to achieve this goal(It seems like "jQuery.ajax" method is a good way, but I am not sure if there is another method can do the same thing, please feel free to leave some better ideas.)
Allow users edit file on the web, this part need to write javascript about loading uploaded file content in ACE editor.
- Jun 24 - 28
Evaluation period. 
Evaluation Features: the website can allow user to upload and list files, users can see their errors returned from ck2cti yaml file on the web
Discuss with mentors about how to deal with syntax errors (highlighting and fix surggestions will be included, but more discussion about details will be helpful, like what suggestions to make for different errors, is there any other functions could be useful). Start diagnosing syntax error, I tend to start with duplicated lines first and missing 'E' error
- Jul 1-5
Diagnose and visualize missing digits from end of line and other syntax error (if found).
- Jul 8-12
Start working on Chemical error(return errors and fix suggestions), I would like to choose  'exceeding collision limit' as first one because I guess this part of code can be similar to [debugging script ](https://github.com/Cantera/ncm-2017-materials/blob/master/mech_debug/thermo_debugging.ipynb)on catera github.
- Jul 15-19
Cooperate with mentors solving discontinuous polynomials and unconsistent thermo for explicit reverse rates 
- Jul 22-26
Evaluation period. 
Evaluation Features: I am not sure about how long will it take for achieving chemical error diagnosis, so I would make a goal for diagnosing all the found syntax error at least
Discuss with mentors about the improvement and plan for next phase and keep working on chemical error diagnosis if not finished. Starting working on auto-correction for syntax error chemical error if diagnosis is finished.
- Jul 29- Aug 2
Working on simplify CVODES error message and dead-end pathways
- Aug 5-9
Keep working if above works are not completed, doing some improvement to Website UI if in feature
- Aug 12- 16
Keep working if above works are not completed, doing some improvement to Website UI if in feature
- Aug 19- Aug 23
Final evaluation period:
Evaluation Features: a web-based application help user to detect either syntax errors or chemical errors and suggest fixes for Cantera input file.

**Comment by bryanwweber** [comment]
Hi @12Chao Thanks for the update, I think this is a good place to go from now. Let's keep updating this issue as the schedule is updated in the coming weeks.

With the first evaluation next week, I'd like to set up a 15-30 minute video call for Friday (6/21) or Monday (6/24) so you can give us a demo of the site and the things you've completed so far. What is your schedule for those days @kyleniemeyer @decaluwe @rwest ?

**Comment by rwest** [comment]
Free Friday afternoon. At at conference Monday.

**Comment by kyleniemeyer** [comment]
I'm good all Friday (pacific time)

**Comment by 12Chao** [comment]
I'm available both two days

**Comment by 12Chao** [comment]
I'm available on both days.

**Comment by decaluwe** [comment]
I'm also available both days. We have the NF call Friday morning (10 am
eastern), right?

**Comment by bryanwweber** [comment]
Yes, the NF call is 10 AM EDT. Let's say 1PM EDT on Friday? We can use my WebEx: https://uconn-cmr.webex.com/meet/bww09001 or another service of your choosing

**Comment by decaluwe** [comment]
Okay, so my schedule changed late yesterday, and I'm only available *until*
1 pm tomorrow - sorry.  I think it's okay if you all want to go ahead
without me, though.

Steven

**Comment by kyleniemeyer** [comment]
OK, I can join at that time on Friday.
